File: framework/languages/java/e2e/RegistrationCodec/Client/src/main/java/systems/zlink/e2e/registrationcodec/client/Scenarios/RcB2ProtobufCodecScenario.java
package systems.zlink.e2e.registrationcodec.client.Scenarios;

import com.google.protobuf.StringValue;
import java.time.Duration;
import systems.zlink.e2e.registrationcodec.client.Support.ScenarioAssert;
import systems.zlink.e2e.registrationcodec.client.Support.ScenarioContext;
import systems.zlink.e2e.registrationcodec.shared.Contracts;

public final class RcB2ProtobufCodecScenario {
    private static final Duration REQUEST_TIMEOUT = Duration.ofSeconds(5);

    private RcB2ProtobufCodecScenario() {
    }

    public static void run(ScenarioContext context) {
        StringValue protobufReply = context.client().requestToChannel(
                Contracts.CHANNEL,
                StringValue.of("protobuf-request"))
            .packetName("ProtobufEcho")
            .timeout(REQUEST_TIMEOUT)
            .await(StringValue.class);
        ScenarioAssert.ensure("echo:protobuf-request".equals(protobufReply.getValue()),
            "RC-B2 request mismatch");
        context.client().sendToChannel(Contracts.CHANNEL, StringValue.of("protobuf-send"))
            .packetName("ProtobufEcho")
            .await();
        ScenarioAssert.waitForEvidence(context.evidence(), "Send", "ProtobufEcho", "protobuf-send");
        System.out.println("scenario RC-B2 passed");
    }
}

File: framework/languages/java/e2e-kotlin/RegistrationCodec/Client/src/main/kotlin/systems/zlink/e2e/kotlin/registrationcodec/client/Scenarios/RcB2ProtobufCodecScenario.kt
package systems.zlink.e2e.kotlin.registrationcodec.client.scenarios

import com.google.protobuf.StringValue
import systems.zlink.e2e.kotlin.registrationcodec.Contracts
import systems.zlink.e2e.kotlin.registrationcodec.client.support.CodecScenarioResult
import systems.zlink.e2e.kotlin.registrationcodec.client.support.EvidenceText
import systems.zlink.e2e.kotlin.registrationcodec.client.support.ScenarioAssert
import systems.zlink.framework.channels.ZLinkClient

class RcB2ProtobufCodecScenario(
    private val client: ZLinkClient,
    private val evidence: EvidenceText,
    private val assert: ScenarioAssert,
) {
    fun run(): CodecScenarioResult {
        val protobufReply = client.requestToChannel(Contracts.CHANNEL, StringValue.of("protobuf-request"))
            .packetName("ProtobufEcho")
            .timeout(requestTimeout)
            .await(StringValue::class.java)
        assert.that(protobufReply.value == "echo:protobuf-request", "RC-B2 request mismatch")
        client.sendToChannel(Contracts.CHANNEL, StringValue.of("protobuf-send"))
            .packetName("ProtobufEcho")
            .await()
        evidence.waitForEvidence("Send", "ProtobufEcho", "protobuf-send")
        println("scenario RC-B2 passed")
        return CodecScenarioResult("RC-B2", "ProtobufEcho", protobufReply.value)
    }
}
